File: activity.py
"""Local activity records, never an authorization source or a broker connection."""
import json
import hashlib
import re
from datetime import datetime, timezone
from pathlib import Path

import device_auth
import workspace

ORDER_FIELDS = ('ticker', 'isin', 'side', 'qty', 'currency', 'limit_price')
RESULT_FIELDS = ('ticker', 'side', 'qty', 'status', 'filled', 'note')
PHASES = {'requested', 'completed', 'not_submitted', 'uncertain'}


def stamp():
    return datetime.now(timezone.utc).isoformat()


def fields(rows, keys):
    if not isinstance(rows, list) or any(not isinstance(row, dict) for row in rows):
        raise ValueError('Invalid activity rows.')
    return [{k: row[k] for k in keys if k in row} for row in rows]


def begin(root, manifest, method):
    """Persist the request before execution. It does not prove broker acceptance."""
    rid = manifest['review_id']
    if not re.fullmatch(r'[a-f0-9]{32}', rid):
        raise ValueError('Invalid activity identifier.')
    record = dict(version=1, id=rid, kind='submission', created_at=stamp(),
                  phase='requested', account=manifest['account'],
                  broker_account=manifest['broker_account'], method=method,
                  orders=fields(manifest['orders'], ORDER_FIELDS), results=[])
    with workspace.LOCK:
        folder = Path(root)/'.workspace'/'activity'
        folder.mkdir(parents=True, exist_ok=True)
        # Preserve the previous version's last receipt before execution replaces it.
        previous = Path(root)/f"orders_result.{manifest['account']}.json"
        if previous.exists():
            try:
                results = fields(read_json(previous), RESULT_FIELDS)
                known = []
                for saved in folder.glob('*.json'):
                    try:
                        old_record = read_json(saved)
                        if old_record.get('account') == manifest['account']:
                            known.append(old_record.get('results'))
                    except (OSError, ValueError, AttributeError): pass
                if results and results not in known:
                    date = datetime.fromtimestamp(previous.stat().st_mtime, timezone.utc).isoformat()
                    old_id = hashlib.sha256(json.dumps([manifest['account'], date, results], sort_keys=True).encode()).hexdigest()[:32]
                    old = dict(version=1, id=old_id, kind='receipt', created_at=date,
                        phase='completed', account=manifest['account'], estimated_date=True, orders=[], results=results)
                    device_auth.atomic_json(folder/(old_id+'.json'), old)
            except (ValueError, TypeError, AttributeError):
                # A corrupt old receipt cannot be represented as a successful trade.
                # Keep its bytes in place by refusing to overwrite it with a new run.
                raise ValueError('The previous receipt could not be archived.')
        path = folder/(rid+'.json')
        if path.exists(): raise ValueError('This submission is already recorded.')
        device_auth.atomic_json(path, record)
    return record


def finish(root, record, ok, not_submitted, results):
    record = dict(record, phase='not_submitted' if not_submitted else 'completed' if ok else 'uncertain',
                  updated_at=stamp(), results=fields(results, RESULT_FIELDS))
    with workspace.LOCK:
        device_auth.atomic_json(Path(root)/'.workspace'/'activity'/(record['id']+'.json'), record)


def read_json(path):
    return json.loads(path.read_text())


def timestamp(value):
    if not isinstance(value, str): raise ValueError('Missing date')
    date = datetime.fromisoformat(value.replace('Z', '+00:00'))
    if date.tzinfo is None: date = date.replace(tzinfo=timezone.utc)
    return date.timestamp()


def listing(root, mode='all', limit=100, offset=0):
    if mode not in ('all', 'personal', 'paper', 'live'):
        raise ValueError('Choose all, personal, paper or live activity.')
    if type(limit) is not int or not 1 <= limit <= 100 or type(offset) is not int or offset < 0:
        raise ValueError('Invalid activity page.')
    root = Path(root)
    entries, warnings = [], []
    with workspace.LOCK:
        folder = root/'.workspace'/'activity'
        for path in sorted(folder.glob('*.json')):
            try:
                if path.is_symlink(): raise ValueError('Linked record')
                r = read_json(path)
                if (r['version'] != 1 or r['kind'] not in ('submission', 'receipt') or r['phase'] not in PHASES
                    or r['account'] not in ('paper', 'live') or r['id'] != path.stem):
                    raise ValueError('Invalid record')
                if r['kind'] == 'submission' and (r.get('method') not in ('pin', 'device')
                    or not isinstance(r.get('broker_account'), str) or not r['broker_account']):
                    raise ValueError('Invalid submission identity')
                timestamp(r['created_at'])
                # Explicit projection: activity never exposes credentials or approval data.
                entries.append(dict(id=r['id'], kind=r['kind'], created_at=r['created_at'], estimated_date=bool(r.get('estimated_date')),
                    updated_at=r.get('updated_at'), account=r['account'], phase=r['phase'],
                    broker_account=r.get('broker_account'), method=r.get('method'),
                    orders=fields(r['orders'], ORDER_FIELDS), results=fields(r['results'], RESULT_FIELDS)))
            except (OSError, ValueError, TypeError, KeyError):
                warnings.append('A saved submission record could not be read. Your history may be incomplete.')
        for name, kind, date_key in [('profile', 'space', 'created_at'), ('portfolio', 'portfolio', 'saved_at'), ('xray', 'xray', None)]:
            path = root/'.workspace'/(name+'.json')
            if not path.exists(): continue
            try:
                data = read_json(path)
                if not isinstance(data, dict): raise ValueError('Invalid saved work')
                date = data.get(date_key) if date_key else None
                estimated = not date
                date = date or datetime.fromtimestamp(path.stat().st_mtime, timezone.utc).isoformat()
                timestamp(date)
                entries.append(dict(id=kind, kind=kind, account='personal', created_at=date, estimated_date=estimated))
            except (OSError, ValueError, TypeError):
                warnings.append('Some saved personal work could not be read.')
        # Earlier versions kept only one receipt per mode. Never invent earlier trades,
        # or associate that receipt with a possibly newer approval/account file.
        for account in ('paper', 'live'):
            path = root/f'orders_result.{account}.json'
            if not path.exists(): continue
            try:
                results = fields(read_json(path), RESULT_FIELDS)
                if results and not any(e.get('results') == results and e['account'] == account for e in entries):
                    entries.append(dict(id='legacy-'+account, kind='receipt', account=account,
                        created_at=datetime.fromtimestamp(path.stat().st_mtime, timezone.utc).isoformat(),
                        estimated_date=True, results=results, orders=[]))
            except (OSError, ValueError, TypeError):
                warnings.append(f'The latest {account} receipt could not be read.')
    entries.sort(key=lambda e: (timestamp(e['created_at']), e['id']), reverse=True)
    selected = [e for e in entries if mode == 'all' or e['account'] == mode]
    return dict(ok=True, entries=selected[offset:offset+limit], total=len(selected),
                next_offset=offset+limit if len(selected) > offset+limit else None,
                warnings=sorted(set(warnings)))

File: activity.js
/* Read-only local history. No broker, approval or submission endpoints. */
const activityNode=id=>document.getElementById(id);
const activityTitles={space:'Your space created',portfolio:'Latest portfolio saved',xray:'Latest Portfolio X-Ray saved',receipt:'Earlier saved order receipt'};
let activityBusy=false,activityOffset=null,activityGeneration=0;
function activityElement(tag,text,className){const e=document.createElement(tag);if(text!==undefined)e.textContent=text;if(className)e.className=className;return e}
function activityDate(value){const d=new Date(value);return Number.isNaN(d.getTime())?'Date unavailable':d.toLocaleString(undefined,{dateStyle:'medium',timeStyle:'short'})}
function activityCard(entry){
 const card=activityElement('article',undefined,'activity-card');
 const meta=activityElement('p',undefined,'activity-meta');
 meta.append(activityElement('span',entry.account==='personal'?'Personal':entry.account==='live'?'Live account':'Paper account','account-tag '+entry.account),activityElement('time',activityDate(entry.created_at)));
 card.append(meta);
 const titles={requested:'Submission requested · check outcome',completed:'Submission results recorded',not_submitted:'Stopped before submission',uncertain:'Submission outcome needs checking'};
 card.append(activityElement('h2',activityTitles[entry.kind]||titles[entry.phase]||'Saved activity'));
 if(entry.kind==='submission'){
  const text={requested:'Lens recorded your approval, but has no saved final outcome for this attempt. Check the broker before creating another plan.',completed:'These are the results saved when Lens finished its submission attempt. Working or partial orders may have changed since then.',not_submitted:'The final checks stopped this attempt before orders were sent.',uncertain:'The attempt ended without a confirmed complete outcome. Some orders may have reached the broker. Check IBKR before submitting again.'};
  card.append(activityElement('p',text[entry.phase]));
  card.append(activityElement('p',(entry.broker_account?'Broker account '+entry.broker_account+' · ':'')+(entry.method==='device'?'Approved with device verification':'Approved with Lens PIN'),'small'));
 }
 if(entry.kind==='receipt')card.append(activityElement('p','A receipt retained from earlier Lens activity. Its original approval method and exact account identifier are not recorded here.'));
 if(entry.estimated_date)card.append(activityElement('p','Date shown is when this file was last saved.','small'));
 if(['portfolio','xray'].includes(entry.kind))card.append(activityElement('p','This is your latest saved version; earlier edits were not recorded.'));
 if(entry.kind==='space')card.append(activityElement('p','Your personal space was created in this Lens installation.'));
 if((entry.orders||[]).length||(entry.results||[]).length){
  const details=activityElement('details');details.append(activityElement('summary','View orders and saved results'));
  if(entry.orders?.length){
   details.append(activityElement('h3','Reviewed orders'));
   const list=activityElement('ul');
   entry.orders.forEach(o=>list.append(activityElement('li',`${o.side} ${o.qty} ${o.ticker}${o.isin?' · '+o.isin:''} · ${o.side==='SELL'?'minimum':'maximum'} ${o.limit_price} ${o.currency||'EUR'} per share`)));details.append(list);
  }
  if(entry.results?.length){
   details.append(activityElement('h3','Results at the time'));
   const list=activityElement('ul');
   entry.results.forEach(r=>list.append(activityElement('li',`${r.side} ${r.qty} ${r.ticker} · ${r.status} · ${r.filled??0} filled${r.note?' · '+r.note:''}`)));details.append(list);
  }else details.append(activityElement('p','No per-order results are saved for this attempt.'));
  card.append(details);
 }
 return card;
}
async function loadActivity(more=false){
 const generation=++activityGeneration;
 activityBusy=true;activityNode('activity-more').disabled=true;
 activityNode('activity-status').textContent='Reading saved activity…';
 if(!more){activityNode('activity-list').replaceChildren();activityOffset=null;activityNode('activity-more').hidden=true}
 try{
  const response=await fetch('/api/activity',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify({account:activityNode('activity-filter').value,limit:30,offset:more?activityOffset:0})});
  const data=await response.json();if(!response.ok||!data.ok)throw Error(data.error||'Activity could not be read. Try again.');
  if(generation!==activityGeneration)return;
  data.entries.forEach(e=>activityNode('activity-list').append(activityCard(e)));
  activityOffset=data.next_offset;
  activityNode('activity-more').hidden=activityOffset===null;
  activityNode('activity-status').textContent=data.total?`${data.total} saved ${data.total===1?'record':'records'}.`:'No saved activity in this view yet.';
  activityNode('activity-warning').textContent=data.warnings.join(' ');
 }catch(e){if(generation===activityGeneration)activityNode('activity-status').textContent=e.message}
 finally{if(generation===activityGeneration){activityBusy=false;activityNode('activity-more').disabled=false}}
}
activityNode('activity-filter').addEventListener('change',()=>loadActivity());
activityNode('activity-refresh').addEventListener('click',()=>loadActivity());
activityNode('activity-more').addEventListener('click',()=>{if(!activityBusy&&activityOffset!==null)loadActivity(true)});
loadActivity();
